### PR TITLE
Improve performance of revision grid

### PR DESCRIPTION
**Pull request by pmiossec**
Improvement over #11441

## Proposed changes

- RevisionGridControl: load revisions by batch (e.g. 1s gain over 2.5s when loading Git repo)
- RevisionReader: cache parent commit hash parsed (less memory usage: between 5 and 10MB on GE repo)
- RevisionGrid: Avoid looking at stash for each revision when possible (no stashes or all already displayed)
- RevisionDataGridView: cache resources and don't calculate same thing twice
- RevisionDataGridView: cache graphic resources between all the cells of the same row
- RevisionGrid): halve the number of short hash calculated on each display

## Screenshots 

no changes

## Test methodology 

- Manual

## Test environment(s) 

- GIT 
- Windows 

## Merge strategy

I agree that the maintainer **rebase** merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitCommands/RevisionReader.cs [hunk 1]
```diff
@@ -240,19 +240,30 @@ private IReadOnlyCollection<GitRevision> GetRevisionsFromArguments(GitArgumentBu
...
                     revisionCount++;
 #endif
-                    subject.OnNext(revision);
+                    revisions.Add(revision);
+                    if (revisions.Count == revisions.Capacity)
+                    {
+                        subject.OnNext(revisions);
+
+                        // ... then use big buffer to load all the remaining revisions with better performance
+                        revisions = new(100_000);
```
Comment (gerhardol) [inline GitCommands/RevisionReader.cs:260]: I suggest slightly smaller chunks (and to clear, there may be more than 100K revisions).
```suggestion
                        if (revisions.Capacity < 10_000)
                        {
                            revisions = new(10_000);
                        }
                        else
                        {
                            revisions.Clear();
                        }
```

File: GitCommands/RevisionReader.cs [hunk 2]
```diff
@@ -240,19 +240,30 @@ private IReadOnlyCollection<GitRevision> GetRevisionsFromArguments(GitArgumentBu
...
 #if TRACE_REVISIONREADER
                     revisionCount++;
 #endif
-                    subject.OnNext(revision);
+                    revisions.Add(revision);
+                    if (revisions.Count == revisions.Capacity)
+                    {
+                        subject.OnNext(revisions);
+
+                        // ... then use big buffer to load all the remaining revisions with better performance
```
Comment (gerhardol) [inline GitCommands/RevisionReader.cs:259]: I am surprised this gives so much improvements.
I tried something similar before optimizing the handling, but it could be different now.
Good finding regardless!

Maybe the reading from stream can be improved too.
Right now the data is reading revision by revision. A bigger buffer could improve (there some improvements in .net8 too).
None of that made any difference when I tested though. (and --log-size  was slower than searching for '0').

File: GitCommands/RevisionReader.cs [hunk 3]
```diff
@@ -281,8 +292,10 @@ private ArgumentBuilder BuildArguments(string revisionFilter, string pathFilter,
             };
         }
 
+        private (ReadOnlyMemory<byte> buffer, ObjectId objectId) _cache = (null, null);
+
         [SuppressMessage("Style", "IDE0057:Use range operator", Justification = "Performance")]
-        private bool TryParseRevision(in ReadOnlySpan<byte> buffer, [NotNullWhen(returnValue: true)] out GitRevision? revision)
+        private bool TryParseRevision(in ReadOnlyMemory<byte> buffer, [NotNullWhen(returnValue: true)] out GitRevision? revision)
```
Comment (gerhardol) [inline GitCommands/RevisionReader.cs:298]: Also surprised that using .Span (cheap though) is more efficient.

File: GitCommands/RevisionReader.cs [hunk 4]
```diff
@@ -301,18 +314,35 @@ private bool TryParseRevision(in ReadOnlySpan<byte> buffer, [NotNullWhen(returnV
...
 
             // The first 40 bytes are the revision ID and the tree ID back to back
-            if (!ObjectId.TryParse(buffer.Slice(0, ObjectId.Sha1CharCount), out ObjectId? objectId) ||
-                !ObjectId.TryParse(buffer.Slice(ObjectId.Sha1CharCount, ObjectId.Sha1CharCount), out ObjectId? treeId))
+            ReadOnlyMemory<byte> commitHash = buffer.Slice(0, ObjectId.Sha1CharCount);
+            ReadOnlySpan<byte> commitHashSpan = commitHash.Span;
+            ObjectId? objectId;
+            if (_cache.objectId != null && commitHashSpan.SequenceEqual(_cache.buffer.Span))
+            {
+                objectId = _cache.objectId;
```
Comment (gerhardol) [inline GitCommands/RevisionReader.cs:322]: ObjectId.TryParse() should be faster with .net9, there could be other improvements too (how to enable vector processing?)
But also nice improvement.

**Review by gerhardol — COMMENTED** [review]
+1 will run and review closer

File: GitCommands/RevisionReader.cs (hunk 1 above)
Comment (mstv) [reply GitCommands/RevisionReader.cs:260]: > I suggest slightly smaller chunks

👍 but independently from the `Capacity`
The grid stays empty until all revisions of the Linux repo have been loaded.
Filters can heavily slow down the loading. Then I want to see every single revision being added in the visible range.

Please be explicit:
`const int minimumInitialCapacity = 100_000;`
How much memory does it cost for a small repo in comparison to `10_000`?

Comment (mstv) [reply GitCommands/RevisionReader.cs]: > > I suggest slightly smaller chunks
> 
> 👍 but independently from the `Capacity`
> **The grid stays empty until all revisions of the Linux repo have been loaded.** Filters can heavily slow down the loading. Then I want to see every single revision being added in the visible range.

TL;DR: At least WorkDir and Index are inserted at the wrong position.

I test with loading all commits. The most recent commit of this repo is from 2023/02/25; the current checkout is from 2023/01/13.
The current **master build** shows a blank grid for more than 30 seconds, too, and traces:

on open with Linux repo

image

on switch to Linux repo (shows the current checkout quickly)

image

and shows WorkDir and Index at the correct position:

image

**This PR shows** a blank grid for more than 30 seconds and traces:

on open with Linux repo

image

on switch to Linux repo (shows the current checkout quickly)

image

but shows WorkDir and Index dozens of pages above the checked-out commit

image

File: GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs [hunk 5]
```diff
@@ -224,106 +232,140 @@ internal void AddColumn(ColumnProvider columnProvider)
...
                 e.Handled = true;
-                _forceRefresh = true;
-                UpdateVisibleRowRange();
+
+                // Ensure triggering is done only one time per row
+                if (e.ColumnIndex == _unhiddableColumnMessageIndex)
+                {
+                    _forceRefresh = true;
+                    UpdateVisibleRowRange();
+                }
```
Comment (mstv) [inline GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs:351]: Good finding!
I had already started to rewrite `UpdateVisibleRowRange` including the graph rendering. And I am going to continue after merging this PR.
But I think here, we should rather let the other column providers properly set `e.Handled = true`. As far as I remember, the graph is the only one which really needs to wait for additional data. The others should be happy with having a non-null revision.

File: GitCommands/RevisionReader.cs [hunk 6]
```diff
@@ -240,19 +240,31 @@ private IReadOnlyCollection<GitRevision> GetRevisionsFromArguments(GitArgumentBu
             using IProcess process = _module.GitCommandRunner.RunDetached(cancellationToken, arguments, redirectOutput: true, outputEncoding: GitModule.LosslessEncoding);
             cancellationToken.ThrowIfCancellationRequested();
 
+            // Initial buffer to give very quick feedback to user
+            List<GitRevision> revisions = new(100);
+            const int commitsBatchSize = 25_000;
```
Comment (mstv) [inline GitCommands/RevisionReader.cs:245]:
```suggestion
            const int firstCommitsBatchSize = 100;
            const int furtherCommitsBatchSize = 25_000;
            List<GitRevision> revisions = new(capacity: initialCommitsBatchSize);
```

File: GitCommands/RevisionReader.cs [hunk 7]
```diff
@@ -240,19 +240,31 @@ private IReadOnlyCollection<GitRevision> GetRevisionsFromArguments(GitArgumentBu
...
                     revisionCount++;
 #endif
-                    subject.OnNext(revision);
+                    revisions.Add(revision);
+                    if (revisions.Count == revisions.Capacity)
+                    {
+                        subject.OnNext(revisions);
+
+                        // ... then use big buffer to load all the remaining revisions with better performance
+                        revisions = new(commitsBatchSize);
```
Comment (mstv) [inline GitCommands/RevisionReader.cs:261]:
```suggestion
                        revisions.Clear();
                        revisions.Capacity = furtherCommitsBatchSize;
```

File: GitCommands/RevisionReader.cs [hunk 8]
```diff
@@ -240,19 +240,31 @@ private IReadOnlyCollection<GitRevision> GetRevisionsFromArguments(GitArgumentBu
...
 
-                if (TryParseRevision(chunk.Span, out GitRevision? revision))
+                if (TryParseRevision(chunk, out GitRevision? revision))
                 {
 #if TRACE_REVISIONREADER
                     revisionCount++;
 #endif
-                    subject.OnNext(revision);
+                    revisions.Add(revision);
+                    if (revisions.Count == revisions.Capacity)
```
Comment (mstv) [inline GitCommands/RevisionReader.cs:256]: In addition, we should use a timer here in case git very slowly yields revisions (e.g. when filtering using "Diff contains").
Can be added in a follow-up.

File: GitUI/UserControls/RevisionGrid/Columns/CommitIdColumnProvider.cs [hunk 9]
```diff
@@ -33,37 +34,59 @@ public override void ApplySettings()
...
-                else if (i > 1)
-                {
-                    _grid.DrawColumnText(e, revision.ObjectId.ToShortString(i - 1), monospaceFont, style.ForeColor, e.CellBounds, useEllipsis: false);
-                }
+            return 0;
+        }
+
+        public override void OnCellPainting(DataGridViewCellPaintingEventArgs e, GitRevision revision, int rowHeight, in CellStyle style)
+        {
+            if (e.FormattedValue == null)
```
Comment (mstv) [inline GitUI/UserControls/RevisionGrid/Columns/CommitIdColumnProvider.cs:63]:
```suggestion
            if (e.FormattedValue is null)
```

File: GitUI/UserControls/RevisionGrid/Columns/CommitIdColumnProvider.cs [hunk 10]
```diff
@@ -33,37 +34,59 @@ public override void ApplySettings()
...
             // Set the grid cell's accessibility text
             if (!revision.IsArtificial)
             {
-                e.Value = revision.ObjectId.ToShortString();
+                if (!_charCount.HasValue)
+                {
+                    _charCount = GetCharLengthForColumnWidth(Column.Width);
+                }
+
+                if (_charCount != 0)
```
Comment (mstv) [inline GitUI/UserControls/RevisionGrid/Columns/CommitIdColumnProvider.cs:86]: Should not matter, but is safer and clearer (to me):
```suggestion
                if (_charCount > 0)
```

File: GitUI/UserControls/RevisionGrid/Graph/RevisionGraphRevision.cs [hunk 11]
```diff
@@ -13,6 +13,8 @@ namespace GitUI.UserControls.RevisionGrid.Graph
     [DebuggerDisplay("{Objectid} {GitRevision.Subject}")]
     public class RevisionGraphRevision
     {
+        // Enough as initial the majority of times because commits nearly never have more than 2 parents.
+        private const int InitialStackSize = 2;
```
Comment (mstv) [inline GitUI/UserControls/RevisionGrid/Graph/RevisionGraphRevision.cs:17]:
```suggestion
        private const int _initialParentStackCapacity = 2;
```

File: GitUI/UserControls/RevisionGrid/Graph/RevisionGraphRevision.cs [hunk 12]
```diff
@@ -78,13 +83,41 @@ public int EnsureScoreIsAbove(int minimalScore)
...
 
-                    DebugHelpers.Assert(parent.Score > revision.Score, "Reorder score failed.");
-
                     maxScore = Math.Max(parent.Score, maxScore);
-                    stack.Push(parent);
+
+                    // Without using a collection (due to performance cost),
+                    // try to queue first parent with a more complex history (i.e. more parents)
+                    // by comparing **only** with previous sibling.
+                    if (previous == null)
```
Comment (mstv) [inline GitUI/UserControls/RevisionGrid/Graph/RevisionGraphRevision.cs:94]:
```suggestion
                    if (previous is null)
```

File: GitUI/UserControls/RevisionGrid/Graph/RevisionGraphRevision.cs [hunk 13]
```diff
@@ -78,13 +83,41 @@ public int EnsureScoreIsAbove(int minimalScore)
...
+                            previous = parent;
+                        }
+                        else
+                        {
+                            stack.Push(parent);
+                        }
+                    }
                 }
+
+                if (previous != null)
```
Comment (mstv) [inline GitUI/UserControls/RevisionGrid/Graph/RevisionGraphRevision.cs:112]:
```suggestion
                if (previous is not null)
```

**Review by mstv — COMMENTED** [review]
basically 👍

File: GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs (hunk 5 above)
Comment (mstv) [reply GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs:351]: I would even set `e.Handled = true` by default before calling `provider.OnCellPainting`. If a providers needs an async update, it shall reset the `e.Handled` to `false`.
The longer I think about it, the less this `_forceRefresh = true` makes sense to me here. There is no reason to invalidate the entire current graph cache.

Comment (mstv) [reply GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs:351]: regarding `e.Handled`: f33646202667a9fc59c57a462f751e65755e3273

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.cs [hunk 14]
```diff
@@ -1223,45 +1228,65 @@ void OnRevisionRead(GitRevision revision)
...
-                    AddArtificialRevisions();
-                }
+                    if (!headIsHandled && (revision.ObjectId.Equals(CurrentCheckout) || CurrentCheckout is null))
+                    {
+                        // Insert artificial worktree/index just before HEAD (CurrentCheckout)
+                        // If grid is filtered and HEAD not visible, insert in OnRevisionReadCompleted()
+                        headIsHandled = true;
+                        _gridView.AddRange(revisionsToDisplay);
+                        revisionsToDisplay.Clear();
+                        AddArtificialRevisions();
```
[the remark this answers is not in the document]
Comment (pmiossec) [reply GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:1285]: I forgot to commit increase of initial capacity. done.

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.cs [hunk 15]
```diff
@@ -1025,56 +1025,61 @@ public void PerformRefreshRevisions(Func<RefsFilter, IReadOnlyList<IGitRef>> get
                     }
                 });
 
-                ThreadHelper.FileAndForget(() =>
+                if (!AppSettings.ShowStashes || getStashRevs.Value.Count == 0)
```
[the remark this answers is not in the document]
Comment (pmiossec) [reply GitUI/UserControls/RevisionGrid/RevisionGridControl.cs]: 👍

File: GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs [hunk 16]
```diff
@@ -224,106 +232,140 @@ internal void AddColumn(ColumnProvider columnProvider)
...
             {
-                Color foreColor = GetForeground(e.State, e.RowIndex);
-                Color commitBodyForeColor = GetCommitBodyForeground(e.State, e.RowIndex);
-                CellStyle cellStyle = new(backBrush, foreColor, commitBodyForeColor, _normalFont, _boldFont, _monospaceFont);
+                return;
+            }
+
+            e.Graphics!.FillRectangle(_cellStyle.Value.BackBrush, e.CellBounds);
+
+            if (_revision is null)
```
[the remark this answers is not in the document]
Comment (pmiossec) [reply GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs:332]: The compiler disagree with you...

File: GitCommands/RevisionReader.cs (hunk 7 above)
Comment (pmiossec) [reply GitCommands/RevisionReader.cs]: We need to create a new collection. I have added a comment.

File: GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs (hunk 16 above)
[the remark this answers is not in the document]
Comment (mstv) [reply GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs:332]: > The compiler disagrees with you...

Because it does not analyze the code that far.
As we are sure, `_revision!` can be used (as in f33646202667a9fc59c57a462f751e65755e3273).

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.cs [hunk 17]
```diff
@@ -1223,45 +1230,65 @@ void OnRevisionRead(GitRevision revision)
                     }
                 }
 
-                if (stashesById is not null)
+                List<GitRevision> revisionsToDisplay = new(revisions.Count + getStashRevs.Value.Count + 2);
```
Comment (mstv) [inline GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:1233]:
```suggestion
                const int artificialCommitCount = 2;
                List<GitRevision> revisionsToDisplay = new(capacity: revisions.Count + getStashRevs.Value.Count + artificialCommitCount);
```

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.cs [hunk 18]
```diff
@@ -1311,8 +1338,7 @@ bool AddArtificialRevisions(IReadOnlyList<ObjectId>? headParents = null)
                 if (headParents is null)
                 {
                     // Add as normal commits at current position
-                    _gridView.Add(workTreeRev);
-                    _gridView.Add(indexRev);
+                    _gridView.AddRange([workTreeRev, indexRev]);
```
Comment (mstv) [inline GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:1341]:
```suggestion
                    revisionsToDisplay.AddRange([workTreeRev, indexRev]);
```

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.cs [hunk 19]
```diff
@@ -1223,45 +1230,65 @@ void OnRevisionRead(GitRevision revision)
...
-                    headIsHandled = true;
-                    AddArtificialRevisions();
-                }
+                    if (!headIsHandled && (revision.ObjectId.Equals(CurrentCheckout) || CurrentCheckout is null))
+                    {
+                        // Insert artificial worktree/index just before HEAD (CurrentCheckout)
+                        // If grid is filtered and HEAD not visible, insert in OnRevisionReadCompleted()
+                        headIsHandled = true;
+                        _gridView.AddRange(revisionsToDisplay);
+                        revisionsToDisplay.Clear();
```
Comment (mstv) [inline GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:1284]: then this is not necessary
```suggestion
```

**Review by mstv — COMMENTED** [review]
in addition to yet open comments, most important: Focused state must affected selected row(s)

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.cs [hunk 20]
```diff
@@ -1223,45 +1228,65 @@ void OnRevisionRead(GitRevision revision)
...
+                                    // Remove current stash displayed and all previous ones (because they are not displayed in grid)
+                                    while (stashesById.Count != 0)
+                                    {
+                                        ObjectId firstStash = stashesById.Keys.First();
+                                        stashesById.Remove(firstStash);
+                                        if (firstStash == stash.ObjectId)
+                                        {
+                                            break;
+                                        }
+                                    }
```
[the remark this answers is not in the document]
Comment (pmiossec) [reply GitUI/UserControls/RevisionGrid/RevisionGridControl.cs]: You're right. I have removed this part for the moment.

**Comment by pmiossec** [comment]
> Commit [3fd13ee](https://github.com/gitextensions/gitextensions/commit/3fd13eed654d3e729c26ba5d28ee98067bb62332) must also take the `Focused` state into account yet.

done.

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.cs (hunk 18 above)
Comment (pmiossec) [reply GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:1334]: I prefer not doing that because otherwise I really don't know how to convert the code of the `else`.

File: GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs [hunk 21]
```diff
@@ -226,104 +233,128 @@ internal void AddColumn(ColumnProvider columnProvider)
             Columns.Add(columnProvider.Column);
         }
 
-        private Color GetForeground(DataGridViewElementStates state, int rowIndex)
+        private Color GetForeground(bool isSelected, bool isNonRelativeGray)
```
Comment (mstv) [inline GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs:236]:
```suggestion
        private Color GetForeground(bool isSelected, bool isFocused, bool isNonRelativeGray)
```

File: GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs [hunk 22]
```diff
@@ -226,104 +233,128 @@ internal void AddColumn(ColumnProvider columnProvider)
...
                 (isNonRelativeGray: true, isSelectedAndFocused: false) => SystemColors.GrayText,
 
                 // (isGray: true, isSelected: true)
-                _ => getHighlightedGrayTextColor()
+                _ => _highlightedGrayTextColor
             };
         }
 
-        private Color GetCommitBodyForeground(DataGridViewElementStates state, int rowIndex)
+        private Color GetCommitBodyForeground(bool isSelected, bool isNonRelativeGray)
```
Comment (mstv) [inline GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs:249]:
```suggestion
        private Color GetCommitBodyForeground(bool isSelected, bool isFocused, bool isNonRelativeGray)
```

**Review by mstv — APPROVED** [review]

**Comment by pmiossec** [comment]
Merging?
